### PR TITLE
Support to disable Checked C type checking

### PR DESCRIPTION
**Pull request by Machiry**
(cc @mwhicks1)
We want to add support to the conversion tool to support partially annotated code. However, partially annotated code could fail type-checking and consequently, the generated AST might not be complete.
For instance, if we want to convert the following code:
```
void foo(_Ptr<int> z) {
  int *x = z;
}
```
clang emits an error:
```
error: initializing 'int *' with an expression of an incompatible type

      '_Ptr<int>'
  int *x = z;
```

And the AST we get is also not complete:
```
./clang -cc1 -ast-dump foo.c
....
`-FunctionDecl 0x10f2fbd0 </home/machiry/checkedc/dd.c:1:1, line:3:1> line:1:6 foo 'void (_Ptr<int>)'
  |-ParmVarDecl 0x10f2faf8 <col:10, col:20> col:20 used z '_Ptr<int>'
  `-CompoundStmt 0x10f301c8 <col:23, line:3:1>
    `-DeclStmt 0x10f301b0 <line:2:3, col:13>
      `-VarDecl 0x10f2fd08 <col:3, col:8> col:8 invalid x 'int *'
1 error generated.
```
More details refer: https://github.com/plum-umd/checkedc-clang/issues/38

The goal of this change is to add support to clang so that all pointers (checked and others) will be treated as non-checked. 
Specifically, we add an option `-fignore-checkedc-pointers`, that will disable CheckedC type checking.

Our conversion tool can use this option to get complete AST, thus making progress in the conversions.

@dtarditi, Please give this a look and let me know if they are fine. I tested these changes with the 
[nocheckedptr.c.zip](https://github.com/microsoft/checkedc-clang/files/4888737/nocheckedptr.c.zip) file.
[](url)

More details on the use case: https://github.com/microsoft/checkedc-clang/pull/847#issuecomment-652375065

### CODE REVIEW

File: clang/lib/Sema/SemaExpr.cpp [hunk 1]
```diff
@@ -8263,22 +8265,29 @@ checkPointerTypesForAssignment(Sema &S, QualType LHSType, QualType RHSType) {
...
   if (!S.Context.pointeeTypesAreAssignable(ltrans, rtrans)) {
      // None of the language extensions below are allowed for pointers
      // that are checked pointers or that contain checked types.
-     if (LHSType->isOrContainsCheckedType() || RHSType->isOrContainsCheckedType())
-         return Sema::Incompatible;
+     if (LHSType->isOrContainsCheckedType() ||
+         RHSType->isOrContainsCheckedType()) {
+       // If ignoring checked pointers is enabled then assignments containing
+       // checked pointers is always compatible
+       return S.getLangOpts().IgnoreCheckedPtr? Sema::Compatible :
```
Comment (mgrang) [inline clang/lib/Sema/SemaExpr.cpp:8288]: Space before `?`

File: clang/lib/Sema/SemaExpr.cpp [hunk 2]
```diff
@@ -8263,22 +8265,29 @@ checkPointerTypesForAssignment(Sema &S, QualType LHSType, QualType RHSType) {
...
   QualType ltrans = QualType(lhptee, 0), rtrans = QualType(rhptee, 0);
   if (!S.Context.pointeeTypesAreAssignable(ltrans, rtrans)) {
      // None of the language extensions below are allowed for pointers
      // that are checked pointers or that contain checked types.
-     if (LHSType->isOrContainsCheckedType() || RHSType->isOrContainsCheckedType())
-         return Sema::Incompatible;
+     if (LHSType->isOrContainsCheckedType() ||
+         RHSType->isOrContainsCheckedType()) {
+       // If ignoring checked pointers is enabled then assignments containing
+       // checked pointers is always compatible
```
Comment (mgrang) [inline clang/lib/Sema/SemaExpr.cpp:8287]: Period needed at the end of comment.

**Comment by Machiry** [comment]
> Overall the changes look fine, if intrusive. One point of concern is how do we ensure developers know/remember to add the `IgnoreCheckedPtr` checks for new Checked C diagnostic code?
> 

That is a good point. Are there any guidelines/design docs for developing Checked C diagnostic code? Currently, how is this ensured for `getLangOpts().CheckedC` checks? We can probably use the same mechanism? Thoughts?

> Also did you get a chance to think about [885f836#diff-0d4e154f927adee7eb243c2df0e59d18R9074](https://github.com/microsoft/checkedc-clang/commit/885f836055ff0a8e8cbf2a8fbbd105c6bfbf1a53#diff-0d4e154f927adee7eb243c2df0e59d18R9074)

We handle this by ignoring the bounds checking errors when `IgnoreCheckedPtr` is enabled.

**Comment by Machiry** [comment]
Thank you for the comments @mgrang . I have fixed them.

File: clang/lib/Sema/SemaExpr.cpp [hunk 3]
```diff
@@ -4786,7 +4786,8 @@ Sema::CreateBuiltinArraySubscriptExpr(Expr *Base, SourceLocation LLoc,
     BaseExpr = LHSExp;
     IndexExpr = RHSExp;
     // Array subscripting not allowed on ptr<T> values
-    if (PTy->getKind() == CheckedPointerKind::Ptr) {
+    if (PTy->getKind() == CheckedPointerKind::Ptr &&
```
[the remark this answers is not in the document]
Comment (Machiry) [reply clang/lib/Sema/SemaExpr.cpp:4789]: Our tool does not do this, but during an interactive conversion session the developer can (by oversight) assigns a `_Ptr` annotation to a pointer used in pointer arithmetic. Ignoring the error here is fine as our tool detects this and tries to fix it in an automated way (i.e., by removing `_Ptr` annotation or making it an `_Array_ptr`).

**Comment by Machiry** [comment]
> @Machiry, thanks for making these changes. Overall this looks good and is the right direction to pursue. I'd like to try to minimize changes. As @mgrang has pointed out, these changes are somewhat intrusive. I have some suggestions for minimizing changes. I'd also like to make it clear that this is a mode specifically designed for use by the Checked C converter, by renaming the flag.
> 

Done.

> Do you expect programmers to turn on this flag and produce programs that run? If not, you should be able to invoke the clang front end in the converter tool without doing code generation. I believe this would allow you to avoid making changes to the "code generation" phases of the compiler.
> 

You are right. We do not expect any programmer to use this to produce programs and yes, we can invoke only the frontend (`-cc1`) of the clang from our tool.  I have removed all my changes to the code generation phase.


> I would like to try to avoid changing the meaning of Checked C extensions, even when this flag is turned on. In practice, I think that means looking at the kind of edits that you might do while incrementally converting a program, and then figuring out whether it makes sense to suppress errors or warnings for that kind of edit. One edit is to convert an unchecked pointer to an checked pointer. That might trigger errors about initialization (although those are usually trivial to address), complaints about missing bounds information, or type checking errors. Another edit is to add a bounds declaration. That could trigger errors or warnings about not being to check bounds declarations or out-of-bounds memory accesses. It would seem that you want those errors. I don't think it makes sense to suppress errors related to checked scopes: there seems to be no need to introduce a checked scope until all the code within the scope uses checked pointers or variables with bounds-safe interface.
> 

You are right. I have minimized my changes by removing almost all changes to `SemaBounds.cpp`.

> Could you add some simple tests where you expect programs to compile without errors when the new compiler flag is turned on?

Done.

File: clang/lib/CodeGen/CGDynamicCheck.cpp [hunk 4]
```diff
@@ -188,7 +188,7 @@ void
 CodeGenFunction::EmitDynamicBoundsCastCheck(const Address BaseAddr,
                                             const BoundsExpr *CastBounds,
                                             const BoundsExpr *SubExprBounds) {
-  if (!getLangOpts().CheckedC)
+  if (!getLangOpts().CheckedC || getLangOpts().IgnoreCheckedPtr)
```
[the remark this answers is not in the document]
Comment (Machiry) [reply clang/lib/CodeGen/CGDynamicCheck.cpp]: Removed all changes to Code-generation.

File: clang/lib/CodeGen/CGDynamicCheck.cpp [hunk 5]
```diff
@@ -108,7 +108,7 @@ void CodeGenFunction::EmitDynamicBoundsCheck(const Address PtrAddr,
                                              const BoundsExpr *Bounds,
                                              BoundsCheckKind CheckKind,
                                              llvm::Value *Val) {
-  if (!getLangOpts().CheckedC)
+  if (!getLangOpts().CheckedC || getLangOpts().IgnoreCheckedPtr)
```
[the remark this answers is not in the document]
Comment (Machiry) [reply clang/lib/CodeGen/CGDynamicCheck.cpp]: Removed all changes to Code-generation.

File: clang/lib/CodeGen/CGDynamicCheck.cpp [hunk 6]
```diff
@@ -96,7 +96,7 @@ void CodeGenFunction::EmitDynamicNonNullCheck(Value *Val,
 void CodeGenFunction::EmitDynamicOverflowCheck(const Address BaseAddr,
                                                const QualType BaseTy,
                                                const Address PtrAddr) {
-  if (!getLangOpts().CheckedC)
+  if (!getLangOpts().CheckedC || getLangOpts().IgnoreCheckedPtr)
```
[the remark this answers is not in the document]
Comment (Machiry) [reply clang/lib/CodeGen/CGDynamicCheck.cpp]: Removed all changes to Code-generation.

File: clang/lib/CodeGen/CGDynamicCheck.cpp [hunk 7]
```diff
@@ -43,7 +43,7 @@ namespace {
 //
 
 void CodeGenFunction::EmitExplicitDynamicCheck(const Expr *Condition) {
-  if (!getLangOpts().CheckedC)
+  if (!getLangOpts().CheckedC || CGM.getLangOpts().IgnoreCheckedPtr)
```
[the remark this answers is not in the document]
Comment (Machiry) [reply clang/lib/CodeGen/CGDynamicCheck.cpp]: Removed all changes to Code-generation.

File: clang/lib/CodeGen/CGDynamicCheck.cpp [hunk 8]
```diff
@@ -59,7 +59,7 @@ void CodeGenFunction::EmitExplicitDynamicCheck(const Expr *Condition) {
 
 static bool shouldEmitNonNullCheck(const CodeGenModule &CGM,
                                    const QualType BaseTy) {
-  if (!CGM.getLangOpts().CheckedC)
+  if (!CGM.getLangOpts().CheckedC || CGM.getLangOpts().IgnoreCheckedPtr)
```
[the remark this answers is not in the document]
Comment (Machiry) [reply clang/lib/CodeGen/CGDynamicCheck.cpp]: Removed all changes to Code-generation.

File: clang/lib/Sema/SemaCast.cpp [hunk 9]
```diff
@@ -2789,7 +2789,7 @@ void CastOperation::CheckCStyleCast(bool IsCheckedScope) {
 
   // Checked C - No C-style casts to unchecked pointer/array type or variadic
   // type in a checked block.
-  if (IsCheckedScope) {
+  if (IsCheckedScope && !Self.getLangOpts().IgnoreCheckedPtr) {
```
[the remark this answers is not in the document]
Comment (Machiry) [reply clang/lib/Sema/SemaCast.cpp]: You are right. This is not a valid case. Removed the check.

File: clang/include/clang/Driver/Options.td [hunk 10]
```diff
@@ -800,7 +800,9 @@ def fbuiltin_module_map : Flag <["-"], "fbuiltin-module-map">, Group<f_Group>,
 def fcheckedc_extension : Flag<["-"], "fcheckedc-extension">, Group<f_Group>, Flags<[CC1Option]>,
   HelpText<"Accept Checked C extension">;
 def fno_checkedc_extension : Flag<["-"], "fno-checkedc-extension">, Group<f_Group>, Flags<[CC1Option]>,
-  HelpText<"Do ont accept Checked C extension">;
+  HelpText<"Do not accept Checked C extension">;
+def fignore_checkedc_pointers : Flag<["-"], "fignore-checkedc-pointers">, Group<f_Group>, Flags<[CC1Option]>,
```
Comment (dtarditi) [inline clang/include/clang/Driver/Options.td:804]: Could you change the name of the command-line flag?

File: clang/lib/AST/ASTContext.cpp [hunk 11]
```diff
@@ -8804,7 +8804,9 @@ QualType ASTContext::mergeFunctionTypes(QualType lhs, QualType rhs,
         // For unchecked return types, a return with
         // bounds is compatible with a return without bounds.
         // The merged type includes the bounds.
-        if (!retType->isUncheckedPointerType())
+        // Ignored the error if we do not want to consider checked pointers.
```
Comment (dtarditi) [inline clang/lib/AST/ASTContext.cpp:8807]: Typl: ignored -> ignore.

File: clang/lib/AST/ASTContext.cpp [hunk 12]
```diff
@@ -8858,7 +8860,9 @@ QualType ASTContext::mergeFunctionTypes(QualType lhs, QualType rhs,
           // For unchecked parameter types, a parameter with
           // bounds is compatible with a parameter without bounds.
           // The merged type includes the bounds.
-          if (!paramType->isUncheckedPointerType())
+          // Ignored the error if we do not want to consider checked pointers.
```
Comment (dtarditi) [inline clang/lib/AST/ASTContext.cpp:8863]: Type: ignored -> ignore.

File: clang/lib/Sema/SemaBounds.cpp [hunk 13]
```diff
@@ -1829,6 +1829,7 @@ namespace {
                                      BoundsExpr *SrcBounds,
                                      EquivExprSets EquivExprs,
                                      CheckedScopeSpecifier CSS) {
+
```
Comment (dtarditi) [inline clang/lib/Sema/SemaBounds.cpp:1832]: Please remove this extra line.

File: clang/lib/Sema/SemaBounds.cpp [hunk 14]
```diff
@@ -1924,6 +1925,7 @@ namespace {
                                   BoundsExpr *ArgBounds,
                                   CheckedScopeSpecifier CSS,
                                   EquivExprSets EquivExprs) {
+
```
Comment (dtarditi) [inline clang/lib/Sema/SemaBounds.cpp:1928]: Please remove this extra line.

File: clang/lib/Sema/SemaBounds.cpp [hunk 15]
```diff
@@ -1951,6 +1953,7 @@ namespace {
                                       BoundsExpr *SrcBounds,
                                       EquivExprSets EquivExprs,
                                       CheckedScopeSpecifier CSS) {
+
```
Comment (dtarditi) [inline clang/lib/Sema/SemaBounds.cpp:1956]: Please remove this extra line.

File: clang/lib/Sema/SemaBounds.cpp [hunk 16]
```diff
@@ -2021,6 +2024,7 @@ namespace {
                                         Expr *Src,
                                         BoundsExpr *SrcBounds,
                                         CheckedScopeSpecifier CSS) {
+
```
Comment (dtarditi) [inline clang/lib/Sema/SemaBounds.cpp:2027]: Please remove this extra line.

File: clang/lib/Sema/SemaExpr.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (dtarditi) [reply clang/lib/Sema/SemaExpr.cpp:4789]: Ok, this makes sense.

File: clang/lib/Sema/SemaExpr.cpp [hunk 17]
```diff
@@ -8263,22 +8265,29 @@ checkPointerTypesForAssignment(Sema &S, QualType LHSType, QualType RHSType) {
...
   QualType ltrans = QualType(lhptee, 0), rtrans = QualType(rhptee, 0);
   if (!S.Context.pointeeTypesAreAssignable(ltrans, rtrans)) {
      // None of the language extensions below are allowed for pointers
      // that are checked pointers or that contain checked types.
-     if (LHSType->isOrContainsCheckedType() || RHSType->isOrContainsCheckedType())
-         return Sema::Incompatible;
+     if (LHSType->isOrContainsCheckedType() ||
+         RHSType->isOrContainsCheckedType()) {
+       // If ignoring checked pointers is enabled then assignments containing
+       // checked pointers is always compatible.
```
Comment (dtarditi) [inline clang/lib/Sema/SemaExpr.cpp:8287]: Typo: is -> are

**Comment by Machiry** [comment]
> Overall, looks great. Please rename the compiler command-line flag too. They were some unnecessary white space changes left behind from reverting some changes also. .

@dtarditi  Thank you for reviewing. Made the necessary changes.

**Comment by dtarditi** [comment]
Hi @Machiry, there are some conflicts that need to be resolved.  Could you resolve them so that I can merge?

Thanks,
David

**Comment by Machiry** [comment]
> Hi @Machiry, there are some conflicts that need to be resolved. Could you resolve them so that I can merge?
> 
> Thanks,
> David

Done.

**Pull request merged**